runtime/elem/builtins/helpers/GainFade.h

```diff
@@ -6,62 +6,116 @@
 namespace elem
 {
 
+    namespace detail {
+        inline double millisecondsToStep(double sampleRate, double ms) {
+            return ms > 1e-6 ? 1.0 / (sampleRate * ms / 1000.0) : 1.0;
+        }
+    }
+
     template <typename FloatType>
     struct GainFade
     {
-        GainFade(double sampleRate, double fadeTimeMs)
-            : step(FloatType(1.0 / (sampleRate * fadeTimeMs / 1000.0)))
+        GainFade(double sampleRate, double fadeInTimeMs, double fadeOutTimeMs, FloatType current = 0.0, FloatType target = 0.0)
+        : currentGain(current)
+        , targetGain(target)
         {
+            setFadeInTimeMs(sampleRate, fadeInTimeMs);
+            setFadeOutTimeMs(sampleRate, fadeOutTimeMs);
         }
 
         GainFade(GainFade const& other)
-            : currentGain(other.currentGain), targetGain(other.targetGain), step(other.step)
+        : currentGain(other.currentGain.load())
+        , targetGain(other.targetGain.load())
+        , step(other.step.load())
+        , inStep(other.inStep)
+        , outStep(other.outStep)
         {
         }
 
         void operator= (GainFade const& other)
         {
-            currentGain = other.currentGain;
-            targetGain = other.targetGain;
-            step = other.step;
+            currentGain.store(other.currentGain.load());
+            targetGain.store(other.targetGain.load());
+            step.store(other.step.load());
+            inStep = other.inStep;
+            outStep = other.outStep;
         }
 
         FloatType operator() (FloatType x) {
-            if (currentGain == targetGain)
-                return (currentGain * x);
+            auto const _currentGain = currentGain.load();
+            if (_currentGain == targetGain.load())
+                return (_currentGain * x);
 
-            auto y = x * currentGain;
-            currentGain = std::clamp(currentGain + step, FloatType(0), FloatType(1));
+            auto y = x * _currentGain;
+            currentGain.store(std::clamp(_currentGain + step.load(), FloatType(0), FloatType(1)));
 
             return y;
         }
 
-        void setTargetGain (FloatType g) {
-            targetGain = g;
+        void process (const FloatType* input, FloatType* output, int numSamples) {
+            auto const _currentGain = currentGain.load();
+            if (_currentGain == targetGain.load()) {
```

Comment:
Let's use the `fpEqual` helper here, and inside the block multiply against the target gain (so that we mul against the target even when the current is just slightly off but still satisfies fpEqual)